solidity/contracts/bridge/RedemptionWatchtower.sol

```diff
@@ -0,0 +1,469 @@
+        uint32 _levelOneDelay,
+        uint32 _levelTwoDelay
+    ) external onlyManager {
+        require(
+            _watchtowerLifetime >= watchtowerLifetime,
+            "New lifetime must not be lesser than current one"
+        );
+
+        // Enforce the 5% hard cap.
+        require(
+            _vetoPenaltyFeeDivisor >= 20 || _vetoPenaltyFeeDivisor == 0,
+            "Redemption veto penalty fee must be in range [0%, 5%]"
+        );
+
+        // Enforce proper relationship between the delay levels. Use
+        // `>=` to allow for setting all delays to zero, if needed.
+        require(
+            _levelTwoDelay >= _levelOneDelay,
+            "Redemption level-two delay must not be lesser than level-one delay"
+        );
+        require(
+            _levelOneDelay >= _defaultDelay,
+            "Redemption level-one delay must not be lesser than default delay"
+        );
+
+        watchtowerLifetime = _watchtowerLifetime;
+        vetoPenaltyFeeDivisor = _vetoPenaltyFeeDivisor;
+        vetoFreezePeriod = _vetoFreezePeriod;
+        defaultDelay = _defaultDelay;
+        levelOneDelay = _levelOneDelay;
+        levelTwoDelay = _levelTwoDelay;
+
+        emit WatchtowerParametersUpdated(
+            _watchtowerLifetime,
+            _vetoPenaltyFeeDivisor,
+            _vetoFreezePeriod,
+            _defaultDelay,
+            _levelOneDelay,
+            _levelTwoDelay
+        );
+    }
+
+    /// @notice Determines whether a redemption request is considered safe.
+    /// @param walletPubKeyHash 20-byte public key hash of the wallet that
+    ///        is meant to handle the redemption request.
+    /// @param redeemerOutputScript The redeemer's length-prefixed output
+    ///        script (P2PKH, P2WPKH, P2SH or P2WSH) that is meant to
+    ///        receive the redeemed amount.
+    /// @param balanceOwner The address of the Bank balance owner whose balance
+    ///        is getting redeemed.
+    /// @param redeemer The address that requested the redemption.
+    /// @return True if the redemption request is safe, false otherwise.
+    ///         The redemption is considered safe when:
+    ///         - The balance owner is not banned,
+    ///         - The redeemer is not banned,
+    ///         - There are no objections against past redemptions from the
+    ///           given wallet to the given redeemer output script.
+    function isSafeRedemption(
+        bytes20 walletPubKeyHash,
+        bytes calldata redeemerOutputScript,
+        address balanceOwner,
+        address redeemer
+    ) external view returns (bool) {
+        if (isBanned[balanceOwner]) {
+            return false;
+        }
+
+        if (isBanned[redeemer]) {
+            return false;
+        }
+
+        uint256 redemptionKey = Redemption.getRedemptionKey(
+            walletPubKeyHash,
+            redeemerOutputScript
+        );
+
+        if (vetoProposals[redemptionKey].objectionsCount > 0) {
```

Comment:
does this not need to be `>= 0`?

Reply:
I don't think so. `vetoProposals[redemptionKey].objectionsCount == 0` is the case for non-objected requests (they don't have the corresponding entry in the `vetoProposals` mapping). In such a case, `isSafeRedemption` must return `true` (assuming the balance owner and redeemer are not banned). If we use `>=0`, `isSafeRedemption` would return `false` for non-objected redemptions which is wrong.

Reply:
ah sorry, I was reading this backwards! :D